### ISSUE TITLE
JoinError from remote s3

### ISSUE THREAD

**Issue by gwaybio** [cytomining/CytoTable#52]
Thanks @d33bs for addressing #47, which may have solved #43, which describes my initial example application (pasted below again)

```python
# Convert CellProfiler SQLite to parquet
cytotable.convert(
    source_path="'s3://cellpainting-gallery/cpg0016-jump/source_1/workspace/backend/Batch1_20221004/UL001643/UL001643.sqlite'",
    dest_path="test.parquet",
    dest_datatype="parquet"
)
```

I received the following error (full traceback bellow)

```python
JoinError: Join failure for task 1 with failed join dependencies from tasks [5]
```

<details>
  <summary>Full python traceback</summary>

```python
---------------------------------------------------------------------------
JoinError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 what = cytotable.convert(
      2     source_path="/".join(manifest_df.sqlite_file[2].split("/")[0:-1]),
      3     dest_path="test.parquet",
      4     dest_datatype="parquet"
      5 )

File ~/miniforge3/envs/jump_sc/lib/python3.10/site-packages/cytotable/convert.py:1186, in convert(source_path, dest_path, dest_datatype, source_datatype, metadata, compartments, identifying_columns, concat, join, joins, chunk_columns, chunk_size, infer_common_schema, drop_null, preset, parsl_config, **kwargs)
   1169 # send sources to be written to parquet if selected
   1170 if dest_datatype == "parquet":
   1171     output = _to_parquet(
   1172         source_path=source_path,
   1173         dest_path=dest_path,
   1174         source_datatype=source_datatype,
   1175         metadata=metadata,
   1176         compartments=compartments,
   1177         identifying_columns=identifying_columns,
   1178         concat=concat,
   1179         join=join,
   1180         joins=joins,
   1181         chunk_columns=chunk_columns,
   1182         chunk_size=chunk_size,
   1183         infer_common_schema=infer_common_schema,
   1184         drop_null=drop_null,
   1185         **kwargs,
-> 1186     ).result()
   1188 return output

File ~/miniforge3/envs/jump_sc/lib/python3.10/concurrent/futures/_base.py:458, in Future.result(self, timeout)
    456     raise CancelledError()
    457 elif self._state == FINISHED:
--> 458     return self.__get_result()
    459 else:
    460     raise TimeoutError()

File ~/miniforge3/envs/jump_sc/lib/python3.10/concurrent/futures/_base.py:403, in Future.__get_result(self)
    401 if self._exception:
    402     try:
--> 403         raise self._exception
    404     finally:
    405         # Break a reference cycle with the exception in self._exception
    406         self = None

File ~/miniforge3/envs/jump_sc/lib/python3.10/site-packages/parsl/dataflow/dflow.py:301, in DataFlowKernel.handle_exec_update(self, task_record, future)
    298     raise RuntimeError("done callback called, despite future not reporting itself as done")
    300 try:
--> 301     res = self._unwrap_remote_exception_wrapper(future)
    303 except Exception as e:
    304     logger.debug("Task {} try {} failed".format(task_id, task_record['try_id']))

File ~/miniforge3/envs/jump_sc/lib/python3.10/site-packages/parsl/dataflow/dflow.py:567, in DataFlowKernel._unwrap_remote_exception_wrapper(future)
    565 result = future.result()
    566 if isinstance(result, RemoteExceptionWrapper):
--> 567     result.reraise()
    568 return result

File ~/miniforge3/envs/jump_sc/lib/python3.10/site-packages/parsl/app/errors.py:122, in RemoteExceptionWrapper.reraise(self)
    118 logger.debug("Reraising exception of type {}".format(self.e_type))
    120 v = self.get_exception()
--> 122 reraise(t, v, v.__traceback__)

File ~/miniforge3/envs/jump_sc/lib/python3.10/site-packages/six.py:719, in reraise(tp, value, tb)
    717     if value.__traceback__ is not tb:
    718         raise value.with_traceback(tb)
--> 719     raise value
    720 finally:
    721     value = None

File ~/miniforge3/envs/jump_sc/lib/python3.10/site-packages/parsl/app/errors.py:160, in wrapper()
    158 from parsl.app.errors import RemoteExceptionWrapper
    159 try:
--> 160     return func(*args, **kwargs)
    161 except Exception:
    162     return RemoteExceptionWrapper(*sys.exc_info())

File ~/miniforge3/envs/jump_sc/lib/python3.10/site-packages/cytotable/convert.py:855, in _to_parquet()
    847 from cytotable.sources import _gather_sources
    849 # gather sources to be processed
    850 sources = _gather_sources(
    851     source_path=source_path,
    852     source_datatype=source_datatype,
    853     targets=list(metadata) + list
(compartments),
    854     **kwargs,
--> 855 ).result()
    857 # if we already have a file in dest_path, remove it
    858 if pathlib.Path(dest_path).is_file():

File ~/miniforge3/envs/jump_sc/lib/python3.10/concurrent/futures/_base.py:458, in result()
    456     raise CancelledError()
    457 elif self._state == FINISHED:
--> 458     return self.__get_result()
    459 else:
    460     raise TimeoutError()

File ~/miniforge3/envs/jump_sc/lib/python3.10/concurrent/futures/_base.py:403, in __get_result()
    401 if self._exception:
    402     try:
--> 403         raise self._exception
    404     finally:
    405         # Break a reference cycle with the exception in self._exception
    406         self = None

JoinError: Join failure for task 1 with failed join dependencies from tasks [5]
```

</details>

parcl log file: [parsl.log](https://github.com/cytomining/CytoTable/files/11492266/parsl.log)

**Comment by d33bs** [comment]
Thanks @gwaybio for raising this! In looking through the logs it looks like DuckDB is having challenges loading the `sqlite_scanner` extension due to what look like possibly OS-level binary read conflicts. DuckDB is used to read source CSV and SQLite data within CytoTable. Doing some investigation here led me to this discussion thread where there is conversation surrounding manual compilations and various options which may resolve the issue https://github.com/duckdb/duckdb/discussions/6307 .

Testing-wise, this error isn't currently caught because my development environment is (coincidentally) compatible with the existing binaries and automated tests run through Debian-based Python Docker images. One possible action item from this issue could be to expand testing into MacOS environments so as to help forecast this or other similar scenarios.

More pragmatically: What version of MacOS are you using? My understanding from https://github.com/duckdb/duckdb/pull/7497 is that MacOS 12.6 is the current minimum compatible version, and that the PR will allow for later DuckDB releases to be compatible with MacOS 11 at a minimum (uncertain of exact release availability). One workaround might involve including this relatively new change from DuckDB somehow.

**Comment by gwaybio** [comment]
Thanks for diagnosing this! 

> What version of MacOS are you using?

I'm on 11.6 (20G165)

My immediate next step is to move this processing to `maple`. I plan on doing this today

**Comment by d33bs** [comment]
Thanks @gwaybio!  Given the new DuckDB change may address this for v11.x on MacOS, it may be worth using [duckdb==0.8.0](https://github.com/duckdb/duckdb/releases/tag/v0.8.0).

I created this branch in an attempt to see what this new version is like here: https://github.com/d33bs/CytoTable/tree/duckdb-version. This appears to break 2 of 17 tests at the moment possibly due to new data typing decisions within DuckDB.

**Comment by gwaybio** [comment]
I tried on maple

OS: Pop!_OS 22.04 LTS 64-bit
CPU: AMD Ryzen 9

and received the same error with a similar log

```python
JoinError: Join failure for task 1 with failed join dependencies from tasks [5]
```

[parsl.log](https://github.com/cytomining/CytoTable/files/11502820/parsl.log)

**Comment by gwaybio** [comment]
update, I'm trying on `maple` with the following environment file, as you suggested in https://github.com/cytomining/CytoTable/issues/52#issuecomment-1552024666

```yaml
name: jump_sc
channels:
  - conda-forge
dependencies:
  - conda-forge::python
  - conda-forge::boto3
  - conda-forge::jupyter
  - conda-forge::jupyterlab
  - conda-forge::ipykernel
  - conda-forge::nb_conda_kernels
  - conda-forge::jupyter_contrib_nbextensions
  - conda-forge::pandas
  - conda-forge::scipy
  - conda-forge::numpy
  - conda-forge::scikit-learn
  - conda-forge::plotnine
  - conda-forge::umap-learn
  - conda-forge::matplotlib
  - pip:
    - git+https://github.com/d33bs/cytotable@c6e0046e76986f9aa837b441186b621fe06b8189
    - duckdb==0.8.0

```

## First attempt

I ran into the same error as before

```python
JoinError: Join failure for task 1 with failed join dependencies from tasks [5]
```

However, the parcl logs told a different story:

```
Traceback (most recent call last):
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/parsl/dataflow/dflow.py", line 301, in handle_exec_update
    res = self._unwrap_remote_exception_wrapper(future)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/parsl/dataflow/dflow.py", line 567, in _unwrap_remote_exception_wrapper
    result.reraise()
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/parsl/app/errors.py", line 122, in reraise
    reraise(t, v, v.__traceback__)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/six.py", line 719, in reraise
    raise value
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/parsl/app/errors.py", line 160, in wrapper
    return func(*args, **kwargs)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/cytotable/sources.py", line 81, in _get_source_filepaths
    if AnyPath(path).is_file()
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/cloudpathlib/s3/s3path.py", line 39, in is_file
    return self.client._is_file_or_dir(self) == "file"
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/cloudpathlib/s3/s3client.py", line 164, in _is_file_or_dir
    return self._s3_file_query(cloud_path)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/cloudpathlib/s3/s3client.py", line 186, in _s3_file_query
    self.client.head_object(
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/client.py", line 530, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/client.py", line 947, in _make_api_call
    http, parsed_response = self._make_request(
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/client.py", line 970, in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/endpoint.py", line 119, in make_request
    return self._send_request(request_dict, operation_model)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/endpoint.py", line 198, in _send_request
    request = self.create_request(request_dict, operation_model)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/endpoint.py", line 134, in create_request
    self._event_emitter.emit(
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/hooks.py", line 412, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/hooks.py", line 256, in emit
    return self._emit(event_name, kwargs)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/hooks.py", line 239, in _emit
    response = handler(**kwargs)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/signers.py", line 105, in handler
    return self.sign(operation_name, request)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/signers.py", line 189, in sign
    auth.add_auth(request)
  File "/home/gway/miniconda3/envs/jump_sc/lib/python3.10/site-packages/botocore/auth.py", line 418, in add_auth
    raise NoCredentialsError()
botocore.exceptions.NoCredentialsError: Unable to locate credentials
```

Presumably, the error occurred because I had not previously configured `maple`.

## Second attempt

I configured aws cli with my credentials and it now seems to be running smoothly. I will report back!

**Comment by d33bs** [comment]
Cheers @gwaybio , thanks for the updates! When working with remote S3 locations, CytoTable relies on the procedures from `cloudpathlib` which stem from `boto3` for S3 access. `Boto3` relies on AWS cli credentials, and I feel you followed the correct procedures to have this updated. Alternatively, for public-facing S3 data an additional parameter may be passed through  `**kwargs` as `cytotable.convert(..., no_sign_request=True)` which is used in a similar way to the [AWS cli option `--no-sign-request`](https://docs.aws.amazon.com/cli/latest/reference/#global-options).

**Comment by gwaybio** [comment]
reopening since there seems to be a task associated with pinning 0.8.0 duckdb and potentially other environment changes to prevent this from occurring again.

**Comment by gwaybio** [comment]
It seems we can close this after merging #60

**Comment by d33bs** [comment]
Marking this as closed with the merge of both #80 and #92 regarding the open item on DuckDB version `>=0.8.0`.
